### ISSUE TITLE
Simplify the GraphStrategy interface

### ISSUE THREAD

**Issue by BrynCooke** [tinkerpop/tinkerpop3#477]
Currently GraphStrategy feels too complicated.

It's because it looks like the formal parameters that should be on the method have been shifted on to the return type.
```
	public UnaryOperator<BiFunction<Direction, String[], Iterator<Vertex>>> getVertexIteratorsVerticesStrategy(
			Context<StrategyWrappedVertex> ctx) {
return lambda crazyness;
}
```
vs
```
	public Iterator<Vertex> getVertexIteratorsVertices(
			Context<StrategyWrappedVertex> ctx, Direction direction, String[] params) {
return an iterator;
}
```

The first example is difficult to document using javadocs and IDEs won't give much assistance to the user. The second example @param can be used for documentation.

If the default implementation must be dynamic then the result could be retained within the context:
```
	public void onVertexIteratorsVertices(
			Context<StrategyWrappedVertex, Iterator<Vertex>> ctx, Direction direction, String[] params) {
ctx.setResult(iterator); //Setting the result indicates that this event has been handled an default processing should not take place.
}

```

**Comment by spmallette** [comment]
The original goal of a `GraphStrategy` was to return a function that was composable to be used in a single `StrategyGraph` wrapper class.  This avoided nesting and tons of boilerplate code.  Your proposed solution doesn't allow for that as far as I can see.

That said - I agree that it is complicated and #476 seeks to undo at least one part of that allowing nesting of `StrategyGraph`.  If we allow nesting then composition isn't really required, thus allowing something like `onVertexIteratorsVertices` to work.

I suspect that we will see some changes here after M7.  thanks.

**Comment by BrynCooke** [comment]
This sounds great
Out of interest will SequenceGraphStrategy be disappearing if nested StrategyGraph is allowed?

Finally just another variation to throw in to the pot:
```
 public Iterator<Vertex> onVertexIteratorsVertices(
            Context<StrategyWrappedVertex, Iterator<Vertex>> ctx, Direction direction, String[] params) {
//Do some stuff
return ctx.proceed(); //Optionally Call the default implementation
//Do stuff 
}
```
Every defender method in the GraphStrategy interface would return ctx.proceed() much in the same way that they all currently return UnaryOperator.identity();

This pattern is similar to the interceptors that you get in DI frameworks or HttpFilters so users will likely have seen it before.

**Comment by spmallette** [comment]
Yes - I think `SequenceGraphStrategy` goes away because it doesn't add anything if nesting works.

**Comment by davidcrouch** [comment]
Developing custom DSLs and extending GraphTraversal<> seemed to be restricted because the base implementations are wrapped in StrategyVertex, StrategyEdge etc. and the current graph strategy functionality does not allow us to extend them unless we replace the complete stack with custom wrappers. 

I am currently working on an experimental re-write of the graph strategy functionality that:-

1) uses the well-known base interfaces as hooks into wrappers to address #477 (Edge, Vertex)  
2) allows developers to inject their own Vertex, Edge implementations (seems to help extending standard DSL and implementing custom DSLs easier - not sure yet).
3) Provides a default implementation to alleviate the need for boilerplate (this is the beauty of the current TP3 strategy implementation over TP2)
4) Uses parts of the wrapper concept in pull request #476. Essentially, it is TP2 wrappers made simpler with Java 8 interface defaults. TP2 wrappers worked - provided a lot of flexibility but came with the overhead of a lot of boilerplate. 
5) Developer still has the ability to easily access/override/provide the underlying methods that do not have defaults (see #490).

My incentive for doing this is that I am trying to develop an authorization strategy (similar to #519) and a history/audit strategy (similar to #520). It seems that having meta-properties on edges would really help developing these but this compromises the data model as per the discussion in #489. So the answer is to provide an abstraction layer where an edge (or to not offend use a similar concept e.g. relationship) can be represented by a vertex in the underlying graph DB. This then drives the need for extending the standard DSLs and/or provide a custom one to integrate the new concept as a 1st class citizen in traversals.

**Comment by davidcrouch** [comment]
I took a stab at another alternate graph strategy design. It's a little quick and dirty (barely compiles) but you can get the idea. I tried to keep it relatively simple and make it behave like the standard interfaces (Graph, Vertex etc.). The "Strategy" interface containing the "factory" methods needs to be modified to fit the Tinkerpop coding style. Please let me know what you think.

https://github.com/davidcrouch/tinkerpop3/tree/custom_strategy_wrappers

**Comment by BrynCooke** [comment]
@davidcrouch
I haven't looked in detail. But as a java dev I would feel much more comfortable with this sort of API.

**Comment by davidcrouch** [comment]
The basic design is this:- For each standard interface (Vertex, Edge, etc.) there is a corresponding XXStrategy interface. These interfaces inherit their standard counterparts plus the Iterator interfaces. They don't add much - a getBaseXX method to get the base element and a getInnerXX method to "peel" back a layer to get the next element e.g.usage:  addVertex(vals) { getInnerGraph.addVertex(vals); .

There is also a "Strategy" interface, which contains methods for instantiating elements  ala Factory pattern. This is to allow the injection of custom elements, which are needed for integrating custom DSLs rather than using "of". (In thinking about it this could also be achieved with the existing design by overriding the methods that return new instances of the "StrategyXX" classes by default and provide an alternate e.g. getGraphAddVertexStrategy.

There is a "DefaultStrategy" implementation, which returns default instances of VertexStrategy, EdgeStrategy etc. I placed all classes in one file for compactness but am not convinced of the idea. 

Lastly, I included an implementation of PartitionStrategy found in /strategy/alternate/util. I again placed all classes in one file as to provide a one-to-one comparison with the existing PartitionStrategy  

One thing about the design is that it would not take much for the functionality in the XXXStrategy methods to be incorporated into the standard interfaces so that extensibility is more of a first class citizen. What this really means is adding the getBaseXX and getInnerXX methods along with the default implementations or changing the existing defaults to use the "getInnerXX" method instead of direct reference.  Just food for thought.
